xdsl/dialects/riscv_ssa.py

```diff
@@ -0,0 +1,711 @@
+        "s1": 9,
+        "a0": 10,
+        "a1": 11,
+        "a2": 12,
+        "a3": 13,
+        "a4": 14,
+        "a5": 15,
+        "a6": 16,
+        "a7": 17,
+        "s2": 18,
+        "s3": 19,
+        "s4": 20,
+        "s5": 21,
+        "s6": 22,
+        "s7": 23,
+        "s8": 24,
+        "s9": 25,
+        "s10": 26,
+        "s11": 27,
+        "t3": 28,
+        "t4": 29,
+        "t5": 30,
+        "t6": 31
+    }
+
+    @staticmethod
+    def from_index(index: int) -> Register:
+        assert 32 > index >= 0
+        register = Register(index)
+        return register
+
+    @staticmethod
+    def from_name(name: str) -> Register:
+        if name in Register.abi_names:
+            return Register.from_index(Register.abi_names[name])
+        if name[0] == 'x' and name[1:].isnumeric():
+            return Register.from_index(int(name[1:]))
+        assert False and "register with unknown name"
+
+    def get_abi_name(self) -> str:
+        for name, index in Register.abi_names.items():
+            if index == self.index:
+                return name
+        assert False and "Register with unknown index"
+
+
+@irdl_attr_definition
+class RegisterAttr(Data[Register]):
+    name = "riscv.reg"
+
+    @staticmethod
+    def parse_parameter(parser: BaseParser) -> Register:
+        token = parser.tokenizer.next_token()
+        if token.text not in Register.abi_names:
+            parser.raise_error("Unknown RISC-V Register {}".format(token.text),
+                               token)
+        return Register.from_name(token.text)
+
+    @staticmethod
+    def print_parameter(data: Register, printer: Printer) -> None:
+        printer.print_string(data.get_abi_name())
+
+    @staticmethod
+    def from_index(index: int) -> RegisterAttr:
+        return RegisterAttr(data=Register.from_index(index))
+
+    @staticmethod
+    def from_name(name: str) -> RegisterAttr:
+        return RegisterAttr(data=Register.from_name(name))
+
+    @staticmethod
+    def from_register(register: Register) -> RegisterAttr:
+        return RegisterAttr(data=register)
+
+
+@irdl_attr_definition
+class LabelAttr(ParametrizedAttribute):
+    name = "riscv.label"
+    label: ParameterDef[StringAttr]
+
+    def string_value(self):
+        return self.label.data
+
+    @staticmethod
+    def from_str(name: str) -> LabelAttr:
+        return LabelAttr([StringAttr.from_str(name)])
+
+
+Op = TypeVar("Op", bound=Operation)
+
+
+@irdl_attr_definition
+class RegisterType(ParametrizedAttribute):
+    name = "riscv_ssa.reg"
+
+
+class Riscv1Rd1Rs1ImmOperation(Operation, ABC):
+    rd: Annotated[OpResult, RegisterType]
+    rs1: Annotated[Operand, RegisterType]
+    immediate: OpAttr[IntAttr | LabelAttr]
+    comment: OptOpAttr[StringAttr]
+
+    @classmethod
+    def get(cls: Type[Op],
+            rs1: Union[Operation, SSAValue],
+            immediate: Union[int, IntAttr, str, LabelAttr],
+            comment: Optional[str] = None) -> Op:
```

Comment:
Does this properly work with type checking? Passing in the type of an op as the first argument and then getting that back?

Reply:
Yep!